wagtail/rich_text/rewriters.py

```diff
@@ -25,101 +27,172 @@ def extract_attrs(attr_string):
+        ):
+            html = (
+                html[: match.start() + offset]
+                + replacement
+                + html[match.end() + offset :]
+            )
+
+            offset += len(replacement) - match.end() + match.start()
+
+        return html
+
+    def extract_tags(self, html):
+        """Helper method to extract and group HTML tags and their attributes.
+
+        Returns the full list of regex matches grouped by tag type as well as
+        the tag attribute dictionaries grouped by tag type.
+        """
+        matches_by_tag_type = defaultdict(list)
+        attrs_by_tag_type = defaultdict(list)
+
+        # Regex used to match <tag ...> tags in the HTML.
+        re_pattern = self.get_opening_tag_regex()
+
+        for match in re_pattern.finditer(html):
+            attrs = extract_attrs(match.group(1))
+            tag_type = self.get_tag_type_from_attrs(attrs)
+
+            matches_by_tag_type[tag_type].append(match)
+            attrs_by_tag_type[tag_type].append(attrs)
+
+        return matches_by_tag_type, attrs_by_tag_type
+
+    def convert_rule_to_bulk_rule(self, rule):
+        def bulk_rule(args):
+            return list(map(rule, args))
+
+        return bulk_rule
 
     def extract_references(self, html):
-        for match in FIND_EMBED_TAG.findall(html):
+        re_pattern = self.get_opening_tag_regex()
+        for match in re_pattern.findall(html):
             attrs = extract_attrs(match)
-            if (
-                "embedtype" not in attrs
-                or attrs["embedtype"] not in self.reference_extractors
-            ):
+            tag_type = self.get_tag_type_from_attrs(attrs)
+
+            if tag_type not in self.reference_extractors:
                 continue
 
-            yield from self.reference_extractors[attrs["embedtype"]](attrs)
+            yield from self.reference_extractors[tag_type](attrs)
 
+        return []
 
-class LinkRewriter:
-    """
-    Rewrites <a linktype="foo"> tags within rich text into the HTML fragment given by the
-    rule for 'foo'. Each link rule is a function that takes a dict of attributes and
-    returns the HTML fragment for the opening tag (only).
-    """
 
-    def __init__(self, link_rules, reference_extractors=None):
-        self.link_rules = link_rules
-        self.reference_extractors = reference_extractors or {}
+class EmbedRewriter(TagRewriter):
+    def get_opening_tag_regex(self):
+        return FIND_EMBED_TAG
+
+    def get_tag_type_from_attrs(self, attrs):
+        return attrs.get("embedtype")
+
+    def get_tag_replacements(self, tag_type, attrs_list):
+        try:
+            rule = self.bulk_rules[tag_type]
+        except KeyError:
+            rule = None
+
+        if not rule:
+            try:
+                rule = self.rules[tag_type]
+            except KeyError:
+                pass
+            else:
+                rule = self.convert_rule_to_bulk_rule(rule)
```

Comment:
@zerolab alternatively we could drop support for non-bulk rules. The Wagtail documentation [describes](https://docs.wagtail.org/en/stable/extending/rich_text_internals.html#rewrite-handlers) implementing custom rewrite handlers that inherit from either EmbedRewriter or LinkRewriter; the ability to define some other rewriter class is technically possible but not really supported. The reason I left the non-bulk functionality in was to avoid making a breaking change - what would you think about doing that? I can certainly add tests but it would be testing a code path that wouldn't be used by the Wagtail code any longer.

Reply:
You are right in that we do say

> New handlers must be Python classes that inherit from either `wagtail.richtext.LinkHandler` or `wagtail.richtext.EmbedHandler`. Your new classes should override at least some of the following methods (listed here for `LinkHandler`, although `EmbedHandler` has an identical signature)


We have two places in which we invoke the the page and embed rewriters:

1. https://github.com/wagtail/wagtail/blob/46cb06a8604b68594757117fb7794fb0145a5e99/wagtail/rich_text/__init__.py#L26 which you've update in this PR.
2. and https://github.com/wagtail/wagtail/blob/46cb06a8604b68594757117fb7794fb0145a5e99/wagtail/admin/rich_text/converters/editor_html.py#L146 -- which needs updating

Usually we follow [a deprecation process](https://docs.wagtail.org/en/stable/contributing/release_process.html#deprecation-policy). While not an official API, I could bet folks have customised or monkey-patched things so that if we made the breaking change, we'd hear. @gasman what do you think?

Reply:
@chosak @zerolab If I'm following correctly, LinkRewriter / EmbedRewriter (note: NOT LinkHandler / EmbedHandler) are entirely undocumented and internal to Wagtail, and so we should feel at liberty to make breaking changes to it without a formal deprecation process - at most we would add a "changes to undocumented internals" release note, if we have reason to believe it's being used in the wild.

However, it sounds like `wagtail.admin.rich_text.converters.editor_html` _is_ still invoking these classes with non-bulk rules. I'm pretty sure that module only exists to support non-Draftail editors such as [wagtail-hallo](https://github.com/wagtail-nest/wagtail-hallo/), and we're probably way overdue for offloading this code out of Wagtail and into the packages in question (if indeed any are still being maintained) - but that's too much scope creep for this PR. So, I'd suggest keeping the `convert_rule_to_bulk_rule` code path in place for the purposes of keeping the editor_html module limping along, and adding the missing test coverage for editor_html to ensure that this code path is tested. It looks like [`EditorHTMLConverter.from_database_format`](https://github.com/wagtail/wagtail/blob/dd48762b188c355a763eeb2a0d6a4c0b0d9fb752/wagtail/admin/rich_text/converters/editor_html.py#L148-L149) is the entry point in question, and the test would look something like the following (in [TestPageLinkHandler](https://github.com/wagtail/wagtail/blob/dd48762b188c355a763eeb2a0d6a4c0b0d9fb752/wagtail/admin/tests/test_rich_text.py#L490)):

```python
    def test_editorhtmlconverter_from_database_format(self):
        events_page_id = Page.objects.get(url_path="/home/events/").pk
        db_html = '<a linktype="page" id="%d">foo</a>' % events_page_id
        converter = EditorHTMLConverter(features=["link"])
        editor_html = converter.from_database_format(db_html)
        self.assertEqual(
            result,
            '<a data-linktype="page" data-id="%d" data-parent-id="2" href="/events/">foo</a>'
            % events_page_id,
        )
```

Reply:
Great suggestion, thanks. Implemented in ab8ba8c.